Reduce fonts footprint from jails

Ashod opened the pull request:
* Fonts are loaded in pre-init, so don't need to be physically available in the jail.
* Reduce systemplate size.

mmeeks commented:
So - its far from clear to me why we don't copy true-type font collections into the jail, but we do copy true type fonts.
Not convinced this is ready.

Ashod commented on kit/Kit.cpp:
```diff
@@ -184,7 +184,8 @@ namespace
                 strcmp(path, "share/config/wizard") != 0 &&
-                strcmp(path, "readmes") != 0;
+                strcmp(path, "readmes") != 0 &&
+                strstr(path, "fonts") == 0;
```
Don't copy `fonts` directories into jail. This leaves out all types of fonts from jails.

Ashod commented on loolwsd-systemplate-setup:
```diff
@@ -105,23 +105,33 @@ done
+    # All the font files are cached in Core at pre-init
+    # from the system, so we don't need them in the jail.
+    # Not having the .ttf files gives us transparent text!
+    find $1 -not -iname "*.ttc" \
+            -not -iname "*.pcf" -not -iname "*.pcf.gz" \
+            -not -iname "*.bdf" -not -iname "*.otf" \
```
This leaves out all but TTF files from *systemplate*. For some reason removing .ttf files from systemplate results in transparent text.

This does *not* mean .ttf files are copied in the jails (see Kit.cpp patch).

Ashod commented:
> So - its far from clear to me why we don't copy true-type font collections into the jail, but we do copy true type fonts.
> Not convinced this is ready.

@mmeeks, as discussed offline: there are two independent changes in this patch, one to setup systemplate and one to setup the jails (Kit.cpp). See my comments above.

But happy to hold this off until we are more confident.

mmeeks commented:
Far from happy here. I'd like to fully understand what's going on in the core first.

Ashod commented:
> Far from happy here. I'd like to fully understand what's going on in the core first.

Sure, sure. This patch is contingent on Core. Sorry, I should've made that explicit. 

My comments are meant to clarify what the patch is actually doing (leaving out all fonts from the jails). But this depends on Core mmapping/preloading them, of course.

Ashod commented:
FWIW, I can't block this from accidentally getting merged because authors can't request changes on their own PR! (Although I can remove it altogether, or change to draft, neither of which is desirable.)

Anyway, this PR is contingent on Core changes!

mmeeks reviewed (changes requested):
Lets request changes then. This patch disturbs me. It removes true-type-font-collections and open-type fonts but not true-type fonts (for no reason I can understand), it drops a chunk of bitmap fonts we would never us - the PCF / PCF.gz / BCF removal is un-controversial and can go in as a standalone commit: please do introduce your function here and bin those as a separate commit =) but lets not drop .ttc or .otf.

Thanks

mmeeks reviewed (changes requested):
I was going to merge this - but it seems we cleanup the un-wanted pcf files anyhow - perhaps we could delete bdf files as an addendum but:

 mkdir -p usr/share || exit 1
-cp -r -p -L /usr/share/fonts usr/share
+copy_fonts_dir /usr/share/fonts .

Why is that a period '.' at the end and not usr/share ? looks like it will move fonts somewhere we don't expect inside the systemplate (and that many not match the fontconfig database (?)

Also - I fear that cp -r --parents -p -L -- will upset FreeBSD / non-gnu copies (or not?).

I'd suggest we drop this pair of patches for now.

Ashod commented:
> I was going to merge this - but it seems we cleanup the un-wanted pcf files anyhow - perhaps we could delete bdf files as an addendum but:

Not sure when pcf's are removed, but I added them and all the other obsolete font files when I found them in my systemplate.

> Why is that a period '.' at the end and not usr/share ? looks like it will move fonts somewhere we don't expect inside the systemplate (and that many not match the fontconfig database (?)

The `cp -r --parents` will recreate the subdirectories, so there is no need to specify anything beyond `.` here.

However, it seems on freebsd we should us -R instead.

I can rework this patch to only filter out the non-ttf fonts, which seems like a decent win of shaving off a few MBs. Also, notice that I had fixed a couple of errors in the old cp commands (perhaps not too visible, but a careful read will reveal them.

mmeeks commented:
Well - for my part - I pretty much hate having the fonts in the system image ;-) it is a really big chunk of 'stuff' and the kernel is inevitably incredibly quicker at copy-on-write page-table / fork optimization magic than we are with a bazillion sys-calls to copy or link this stuff across.

So - if we have time to invest in dev / test here - we should IMHO abandon this - and focus 100% on pre-loading all the fonts and enlarging the CairoFontsCache, and populating it before fork. That would give us simpler code, and avoid needing to get this patch and any associated regressions in & handled =)

Ashod commented:
> Well - for my part - I pretty much hate having the fonts in the system image ;-) 

This patch *removes* the fonts from the system image.

Anyway, abandoned for now.
